src/Data/Array/NonEmpty/Internal.ss

```diff
@@ -0,0 +1,37 @@
+(library (Data.Array.NonEmpty.Internal foreign)
+  (export foldr1Impl
+          foldl1Impl
+          traverse1Impl)
+  (import (only (rnrs base) define lambda cons let if quote)
+          (only (chezscheme) fx- fx< fx>= fx+)
+          (prefix (purs runtime lib) rt:)
+          (prefix (purs runtime srfi :214) srfi:214:))
+
+  (define foldr1Impl
+    (lambda (f xs)
+      (let loop ([acc (srfi:214:flexvector-back xs)]
+                 [i (fx- (rt:array-length xs) 2)])
+        (if (fx>= i 0)
+          (loop ((f (rt:array-ref xs i)) acc) (fx- i 1))
+          acc))))
+
+  (define foldl1Impl
+    (lambda (f xs)
+      (let loop ([acc (rt:array-ref xs 0)] [i 1])
+        (if (fx< i (rt:array-length xs))
+          (loop ((f acc) (rt:array-ref xs i)) (fx+ i 1))
+          acc))))
+
+  (define traverse1Impl
+    (lambda (apply map f)
+
+      (define kons (lambda (x) (lambda (ys) (cons x ys))))
+      (define singleton (lambda (x) (cons x '())))
```

Comment:
I think we could just use the built-in `list` here instead of `singleton`. Or `(define singleton list)`.

Reply:
Ha, I was wondering how can there be no built-in `singleton`, forgetting about `list` completely :D